src/main/java/com/iota/iri/network/NeighborRouter.java

```diff
@@ -0,0 +1,539 @@
+    }
+
+    public boolean addNeighbor(String rawURI) throws IOException {
+        if (connectedNeighbors.size() >= config.getMaxNeighbors()) {
+            return false;
+        }
+        Optional<URI> optUri = parseURI(rawURI);
+        if (!optUri.isPresent()) {
+            return false;
+        }
+        URI neighborURI = optUri.get();
+        InetSocketAddress inetAddr = new InetSocketAddress(neighborURI.getHost(), neighborURI.getPort());
+        if (!initNeighborConnection(neighborURI, inetAddr)) {
+            neighborsToConnectTo.add(neighborURI);
+        }
+        return true;
+    }
+
+    public boolean removeNeighbor(String rawURI) {
+        Optional<URI> optUri = parseURI(rawURI);
+        if (!optUri.isPresent()) {
+            return false;
+        }
+
+        // remove the neighbor from connection attempts.
+        // this is racy when a connection is initialized and key.isConnectable() is ongoing
+        // as the neighbor will be added by it again
+        neighborsToConnectTo.remove(optUri.get());
+
+        URI neighborURI = optUri.get();
+        InetSocketAddress inetAddr = new InetSocketAddress(neighborURI.getHost(), neighborURI.getPort());
+        if (inetAddr.isUnresolved()) {
+            log.warn("unable to remove neighbor {} as IP address couldn't be resolved", rawURI);
+            return false;
+        }
+        String identity = String.format("%s:%d", inetAddr.getAddress().getHostAddress(), inetAddr.getPort());
+        Neighbor neighbor = connectedNeighbors.get(identity);
+        if (neighbor == null) {
+            return false;
+        }
+        // the neighbor will be disconnected inside the selector loop
+        neighbor.setState(NeighborState.MARKED_FOR_DISCONNECT);
+        return true;
+    }
+
+    public static Optional<URI> parseURI(final String uri) {
+        if (uri.isEmpty()) {
+            return Optional.empty();
+        }
+
+        URI neighborURI;
+        try {
+            neighborURI = new URI(uri);
+        } catch (URISyntaxException e) {
+            log.error("URI {} raised URI Syntax Exception", uri);
+            return Optional.empty();
+        }
+        if (!isURIValid(neighborURI)) {
+            return Optional.empty();
+        }
+        return Optional.of(neighborURI);
+    }
+
+    public static boolean isURIValid(final URI uri) {
+        if (uri == null) {
+            log.error("Cannot read URI schema, please check neighbor config!");
+            return false;
+        }
+
+        if (!uri.getScheme().equals("tcp")) {
+            log.error("'{}' is not a valid URI schema", uri);
+            return false;
+        }
+        return true;
+    }
+
+    public TxPipeline getTxPipeline() {
+        return txPipeline;
+    }
+
+    public ConcurrentHashMap<String, Neighbor> getConnectedNeighbors() {
+        return connectedNeighbors;
+    }
+
+    public void gossipTransaction(Neighbor neighbor, TransactionViewModel tvm) throws Exception {
```

Comment:
the name `gossipTransaction` sounds like `broadcast` as if it will gossip to an arbitrary amount of neighbors.
maybe `gossipTransactionTo` or something more personal :)

Reply:
True, `gossipTransactionTo` makes more sense, thanks :)